### ISSUE TITLE
orix.sampling.get_sample_local() is quite memory intensitve

### ISSUE THREAD

**Issue by hakonanes** [pyxem/orix#175]
This code fills my 16 GB memory (from 6 GB used before running) immediately

```python
>>> import numpy as np
>>> from orix import sampling, quaternion
>>> r = quaternion.Rotation.from_euler(np.deg2rad([133.3, 88.7, 177.8]))
>>> r2 = sampling.get_sample_local(resolution=0.5, center=r, grid_width=5)
```

Am I wrong in assuming that this code would give me about 5^3 / 0.5 = 250 rotations about `r`? If I'm right, we should revisit the implementation of `get_sample_local()`.

**Comment by pc494** [comment]
Resolution is a difficult concept here, so I won't speculate about how many rotations you might actually expect. But the current implementation is very naive (it samples all of orientation space then cuts it back down). 

I think I could improve upon this.

**Comment by hakonanes** [comment]
Understandable. But 250 is a reasoanble estimate, give or take, I would assume.

I haven't used this function before now, but I believe it can be usefuly in our work. I would be happy to assist.

**Comment by pc494** [comment]
The basic problems were:

1) Sampling uniformly while retaining a constraint (ie. rather than using a rejection based sampling).
2) Not introducing randomness (eg. Any method that involves a Gaussian)

I'm currently having a look and I think we probably could improve things

**Comment by pc494** [comment]
Okay, I've now got an idea, could probably get this done for the start of next week, would that be okay?

**Comment by hakonanes** [comment]
Yeah, totally, there is no rush on this!

**Comment by hakonanes** [comment]
I believe `Object3d.unique()` is the source for the memory error, which I see by trying out #175 as well (will finish review on Sunday). I think there is room for improvement in that method, I will have a look at that hopefully next week.

**Comment by pc494** [comment]
> I believe `Object3d.unique()` is the source for the memory error, which I see by trying out #175 as well (will finish review on Sunday). I think there is room for improvement in that method, I will have a look at that hopefully next week.

`"""Returns a new object containing only this object's unique entries."""`

Yeah that'll double the usage, but I think #175 should still make it a lot smaller.

**Comment by pc494** [comment]
Closed by #177
